email/exporter.go

```diff
@@ -0,0 +1,169 @@
+package email
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/time/rate"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/letsencrypt/boulder/core"
+	exporterpb "github.com/letsencrypt/boulder/email/proto"
+	berrors "github.com/letsencrypt/boulder/errors"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+const (
+	// Our daily limit is determined by the tier of our Salesforce account. For
+	// more information, see:
+	// https://developer.salesforce.com/docs/marketing/pardot/guide/overview.html?q=rate%20limits
+
+	// ratelimit represents our daily limit of 50,000 requests.
+	rateLimit = 50000.0 / 86400.0
+
+	// numWorkers is the number of concurrent workers processing the email
+	// queue. We also use this as the burst limit for the rate limiter.
+	numWorkers = 5
+
+	// queueCap enforces a maximum stack size to prevent unbounded growth.
+	queueCap = 10000
+)
+
+var ErrQueueFull = errors.New("email export queue is full")
+
+// ExporterImpl implements the gRPC server and processes email exports.
+type ExporterImpl struct {
+	sync.RWMutex
+
+	toSend []string
+	client *PardotClient
+	log    blog.Logger
+}
+
+// NewExporterImpl creates a new ExporterImpl.
+func NewExporterImpl(client *PardotClient, scope prometheus.Registerer, logger blog.Logger) *ExporterImpl {
+	impl := &ExporterImpl{
+		toSend: make([]string, 0, queueCap),
+		client: client,
+		log:    logger,
+	}
+
+	queueGauge := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "email_exporter_queue_length",
+		Help: "Current length of the email export queue",
+	}, func() float64 {
+		impl.RLock()
+		defer impl.RUnlock()
+		return float64(len(impl.toSend))
+	})
+	scope.MustRegister(queueGauge)
+
+	return impl
+}
+
+// UpsertEmails enqueues the provided email addresses. If the queue is near
+// capacity, only enqueues as many emails as can fit. Returns ErrQueueFull if
+// some or all emails were dropped.
+func (impl *ExporterImpl) UpsertEmails(ctx context.Context, req *exporterpb.UpsertEmailsRequest) (*emptypb.Empty, error) {
+	if core.IsAnyNilOrZero(req, req.Emails) {
+		return nil, berrors.InternalServerError("Incomplete UpsertEmails request")
+	}
+
+	impl.Lock()
+	defer impl.Unlock()
+
+	spotsLeft := queueCap - len(impl.toSend)
+	if spotsLeft <= 0 {
+		return nil, ErrQueueFull
+	}
+
+	toAdd := req.Emails
+	if len(toAdd) > spotsLeft {
+		toAdd = toAdd[:spotsLeft]
+	}
+
+	impl.toSend = append(impl.toSend, toAdd...)
+
+	if len(toAdd) < len(req.Emails) {
+		impl.log.Errf("Dropped %d emails due to queue capacity", len(req.Emails)-len(toAdd))
+		return nil, ErrQueueFull
+	}
```

Comment:
I'm a little leery of enqueueing _part_ of the request, then returning ErrQueueFull. Sure, we control the client, but in general I'd expect "queue full" to be a retryable error, in which case the client will retry _all_ of the emails it tried the first time, potentially exacerbating the nearly-full-queue situation.

I'd say just return ErrQueueFull (potentially augmented with the size of the request and the number of slots left, for debugging in the logs) if the queue doesn't have enough room for the whole request:
```suggestion
	spotsLeft := queueCap - len(impl.toSend)
	if spotsLeft < len(req.Emails) {
		return nil, ErrQueueFull
	}

	impl.toSend = append(impl.toSend, req.Emails...)
```